<a href="https://colab.research.google.com/github/sreoshi123/Assignments/blob/main/Tensorflow_Keras_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# 1. In this task, you'll build an ANN and train and test it using the MNIST data. This ANN should consist of two hidden layers and one output layer. All of the hidden layers should be dense. The first layer and the second layer should have neuron sizes of 32 and 16, respectively. Train this model for 20 epochs, and compare your training and test set performance with the example in the lesson. Is there any difference? If so, why?

In [3]:
import warnings
warnings.filterwarnings('ignore')

from tensorflow.keras.datasets import mnist

In [5]:
# Loading the data and preprocessing
(X_train,y_train), (X_test, y_test) = mnist.load_data()

input_dim = 784 # 28*28
output_dim = nb_classes = 10
batch_size = 128
nb_epoch = 20

X_train = X_train.reshape(60000,input_dim)
X_test = X_test.reshape(10000,input_dim)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

In [8]:
# One hot encoding the target variable
from tensorflow.keras.utils import to_categorical

Y_train = to_categorical(y_train, nb_classes)
Y_test = to_categorical(y_test, nb_classes)

In [10]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

In [12]:
model = Sequential()
# our first dense layer
model.add(Dense(32, input_shape=(784,), activation = 'relu'))
# our second dense layer
model.add(Dense(16, activation ='relu'))
# last layer is the output layer
model.add(Dense(10, activation='softmax'))


In [13]:
model.compile(optimizer = 'sgd', loss = 'categorical_crossentropy',metrics = ['accuracy'])

In [14]:
# setting verbose = 1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size = batch_size, epochs = 20, verbose=1)

Epoch 1/20
469/469 [==============================] - 2s 3ms/step - loss: 1.6643 - accuracy: 0.4801
Epoch 2/20
469/469 [==============================] - 2s 5ms/step - loss: 0.6799 - accuracy: 0.8157
Epoch 3/20
469/469 [==============================] - 2s 4ms/step - loss: 0.4624 - accuracy: 0.8721
Epoch 4/20
469/469 [==============================] - 1s 3ms/step - loss: 0.3863 - accuracy: 0.8914
Epoch 5/20
469/469 [==============================] - 1s 3ms/step - loss: 0.3478 - accuracy: 0.9018
Epoch 6/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3234 - accuracy: 0.9084
Epoch 7/20
469/469 [==============================] - 1s 3ms/step - loss: 0.3054 - accuracy: 0.9135
Epoch 8/20
469/469 [==============================] - 1s 3ms/step - loss: 0.2909 - accuracy: 0.9172
Epoch 9/20
469/469 [==============================] - 1s 3ms/step - loss: 0.2783 - accuracy: 0.9215
Epoch 10/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2675 - accuracy: 0.9241

In [15]:
score = model.evaluate(X_test, Y_test, verbose =0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.19683483242988586
Test accuracy: 0.9448000192642212


The model achieved 94% accuracy in both training and test set. This level is lower than that of the example in the checkpoint. Since the layers include less number of neurons in this model, the model is simpler than the one of the example.

Q2.) You'll also build an ANN in this task. This time, this ANN should have 5 hidden layers and 1 output layer. All the layers should be dense. The neuron numbers for the hidden layers should be 1024, 512, 256, 128 and 64. Train this model 20 epochs and test it using the same data from the previous task and compare your results. Is there any difference? If so, why?

In [16]:
model = Sequential()
# our first dense layer
model.add(Dense(1024,input_shape = (784,), activation = 'relu'))
# our second dense layer
model.add(Dense(512, activation = 'relu'))
# our third dense layer
model.add(Dense(256,activation ='relu'))
# our fourth dense layer
model.add(Dense(128,activation='relu'))
# our fifth dense layer
model.add(Dense(64,activation='relu'))
# last layer is the output layer
model.add(Dense(10,activation='softmax'))

In [17]:
model.compile(optimizer = 'sgd',loss='categorical_crossentropy',metrics = ['accuracy'])

In [18]:
# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size = batch_size, epochs =20, verbose =1)

Epoch 1/20
469/469 [==============================] - 16s 32ms/step - loss: 1.1887 - accuracy: 0.7029
Epoch 2/20
469/469 [==============================] - 15s 33ms/step - loss: 0.3681 - accuracy: 0.8985
Epoch 3/20
469/469 [==============================] - 15s 31ms/step - loss: 0.2817 - accuracy: 0.9203
Epoch 4/20
469/469 [==============================] - 15s 32ms/step - loss: 0.2360 - accuracy: 0.9326
Epoch 5/20
469/469 [==============================] - 15s 31ms/step - loss: 0.2033 - accuracy: 0.9413
Epoch 6/20
469/469 [==============================] - 15s 31ms/step - loss: 0.1783 - accuracy: 0.9488
Epoch 7/20
469/469 [==============================] - 14s 31ms/step - loss: 0.1583 - accuracy: 0.9545
Epoch 8/20
469/469 [==============================] - 14s 31ms/step - loss: 0.1422 - accuracy: 0.9591
Epoch 9/20
469/469 [==============================] - 14s 30ms/step - loss: 0.1281 - accuracy: 0.9630
Epoch 10/20
469/469 [==============================] - 14s 30ms/step - loss: 0.116

In [19]:
score = model.evaluate(X_test, Y_test, verbose =0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.09369824081659317
Test accuracy: 0.9724000096321106


In this case the model achieved almost 98% accuracy in the training set and 97% accuracy in the test set. The model here is more complex than the model of the previous task and the model in the exanple. Because this model is more complex than the previous two models it achieved higher performace in the traning set. However, the difference between the training and test score widened a little bit. It may be a sign that this model is too complex and it started to overfit.